networkx/algorithms/approximation/steinertree.py

```diff
@@ -114,18 +114,29 @@ def _kou_steiner_tree(G, terminal_nodes, weight):
 
 def _remove_nonterminal_leaves(G, terminals):
     terminals_set = set(terminals)
-    degree_one_nodes = {n for n in G if G.degree(n) == 1}
-    nonterminal_leaves = degree_one_nodes - terminals_set
+    self_looped_nodes = {n for n in G if n in G[n] and len(set(G[n])) == 2}
+    degree_1_not_looped = {n for n in G if (G.degree(n) == 1) and (n not in G[n])}
+    nonterminal_leaves = (self_looped_nodes | degree_1_not_looped) - terminals_set
+
     while nonterminal_leaves:
         possible_nonterminal_leaves = set()
         for n in nonterminal_leaves:
             possible_nonterminal_leaves = possible_nonterminal_leaves | set(
                 G.neighbors(n)
-            )
+            ) - {n}
```

Comment:
While I know the existing code makes the code a little ugly, while we're here can we collapse this to something like:
```python
nbrs = ([nbr for nbr in G.neighbors(n) if n != nbr] for n in nonterminal_leaves)
possible_nonterminal_leaves = set().union(*nbrs) - terminals_set
```

Reply:
I think the previous code was a little easier to read, but I'm sure that purely subjective. 

I've implemented your suggestion.